File: idea-plugin/src/main/java/com/zzf/rikki/idea/agent/tools/LiteIdeTools.java
package com.zzf.rikki.idea.agent.tools;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.intellij.openapi.project.Project;
import com.zzf.rikki.idea.IdeBridgeServer;

import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;

public class LiteIdeTools {
    private final Project project;
    private final ObjectMapper mapper;
    private volatile JsonNode ideContextNode;
    private volatile CapabilitySnapshot cachedCapabilitySnapshot = new CapabilitySnapshot(false, List.of());

    public LiteIdeTools(Project project, ObjectMapper mapper) {
        this.project = project;
        this.mapper = mapper;
        this.ideContextNode = mapper.createObjectNode();
    }

    public CapabilitySnapshot capabilitySnapshot() {
        return cachedCapabilitySnapshot;
    }

    public CapabilitySnapshot refreshCapabilities() {
        JsonNode node = callBridgeJson(mapper.createObjectNode().put("action", "capabilities"));
        CapabilitySnapshot snapshot = parseCapabilitySnapshot(node);
        cachedCapabilitySnapshot = snapshot;
        return snapshot;
    }

    public String context(JsonNode args) throws Exception {
        JsonNode ctx = ideContextNode;
        if (ctx == null || ctx.isNull() || ctx.isMissingNode() || ctx.size() == 0) {
            return "No IDE context is available. Ensure the plugin is loaded and project is indexed.";
        }
        String query = args.path("query").asText("all");
        List<String> keys = new ArrayList<>();
        if (args.has("keys") && args.path("keys").isArray()) {
            args.path("keys").forEach(node -> {
                String key = node.asText("");
                if (!key.isBlank()) {
                    keys.add(key);
                }
            });
        }
        ObjectNode filtered = mapper.createObjectNode();
        ctx.fields().forEachRemaining(entry -> {
            String key = entry.getKey();
            if ((!keys.isEmpty() && keys.contains(key)) || (keys.isEmpty() && matchesQuery(key, query))) {
                filtered.set(key, entry.getValue());
            }
        });
        return mapper.writerWithDefaultPrettyPrinter().writeValueAsString(filtered);
    }

    public String action(JsonNode args) throws Exception {
        String operationRaw = firstNonBlank(args.path("operation").asText(""), args.path("action").asText("")).toLowerCase(Locale.ROOT);
        if (operationRaw.isBlank()) {
            return prettyJson(errorNode("missing_operation", "Field 'operation' is required for ide_action."));
        }
        if ("capabilities".equals(operationRaw)) {
            return capabilities();
        }
        if ("build".equals(operationRaw)) {
            return prettyJson(unsupportedBuildNode());
        }
        CapabilitySnapshot snapshot = refreshCapabilities();
        if (!snapshot.getBridgeAvailable()) {
            return prettyJson(errorNode("bridge_unavailable", "IDE bridge unavailable. Use bash for build/test operations."));
        }
        if (!snapshot.getActionOperations().contains(operationRaw)) {
            return prettyJson(errorNode("unsupported_operation", "Unsupported ide_action operation: " + operationRaw));
        }
        ObjectNode payload = mapper.createObjectNode();
        switch (operationRaw) {
            case "run":
            case "test":
                payload.put("action", "start");
                payload.put("operation", operationRaw);
                break;
            case "status":
                payload.put("action", "status");
                break;
            case "cancel":
                payload.put("action", "cancel");
                break;
            default:
                return prettyJson(errorNode("unsupported_operation", "Unsupported ide_action operation: " + operationRaw));
        }
        copyIfPresent(args, payload, "mode", "configuration", "configurationName", "name", "executor", "jobId", "sinceRevision", "waitMs");
        JsonNode firstResponse = callBridgeJson(payload);
        boolean wait = args.path("wait").asBoolean(false);
        if (!wait || (!"run".equals(operationRaw) && !"test".equals(operationRaw))) {
            return prettyJson(firstResponse);
        }
        String jobId = firstNonBlank(firstResponse.path("jobId").asText(""));
        if (!firstResponse.path("ok").asBoolean(false) || jobId.isBlank()) {
            return prettyJson(firstResponse);
        }
        if (!snapshot.getActionOperations().contains("status")) {
            return prettyJson(firstResponse);
        }
        long timeoutMs = Math.max(1_000L, Math.min(900_000L, args.path("timeoutMs").asLong(120_000L)));
        long pollIntervalMs = Math.max(200L, Math.min(10_000L, args.path("pollIntervalMs").asLong(1_000L)));
        long deadline = System.currentTimeMillis() + timeoutMs;
        long sinceRevision = 0L;
        JsonNode latest = firstResponse;
        while (System.currentTimeMillis() < deadline) {
            Thread.sleep(pollIntervalMs);
            ObjectNode statusPayload = mapper.createObjectNode()
                    .put("action", "status")
                    .put("jobId", jobId)
                    .put("sinceRevision", sinceRevision)
                    .put("waitMs", pollIntervalMs);
            latest = callBridgeJson(statusPayload);
            sinceRevision = Math.max(sinceRevision, latest.path("logRevision").asLong(0L));
            if (isTerminalStatus(latest.path("status").asText(""))) {
                return prettyJson(latest);
            }
        }
        ObjectNode timeoutNode = latest instanceof ObjectNode objectNode ? objectNode : mapper.createObjectNode();
        timeoutNode.put("ok", false);
        timeoutNode.put("status", "timeout");
        timeoutNode.put("summary", "Timed out waiting for IDE job completion.");
        timeoutNode.put("output", firstNonBlank(timeoutNode.path("output").asText(""), "Timed out waiting for IDE job completion."));
        return prettyJson(timeoutNode);
    }

    public String capabilities() throws Exception {
        JsonNode node = callBridgeJson(mapper.createObjectNode().put("action", "capabilities"));
        cachedCapabilitySnapshot = parseCapabilitySnapshot(node);
        return prettyJson(node);
    }

    public void setIdeContextNode(JsonNode ideContextNode) {
        this.ideContextNode = ideContextNode == null ? mapper.createObjectNode() : ideContextNode;
    }

    private CapabilitySnapshot parseCapabilitySnapshot(JsonNode node) {
        if (!node.path("ok").asBoolean(false)) {
            return new CapabilitySnapshot(false, List.of());
        }
        LinkedHashSet<String> operations = new LinkedHashSet<>();
        collectOps(node.path("asyncOperations"), operations);
        collectOps(node.path("directOperations"), operations);
        return new CapabilitySnapshot(true, new ArrayList<>(operations));
    }

    private void collectOps(JsonNode source, LinkedHashSet<String> dest) {
        if (!source.isArray()) {
            return;
        }
        for (JsonNode node : source) {
            String operation = node.asText("").trim().toLowerCase(Locale.ROOT);
            if (List.of("run", "test", "status", "cancel", "capabilities").contains(operation)) {
                dest.add(operation);
            }
        }
    }

    private boolean matchesQuery(String key, String query) {
        if ("all".equalsIgnoreCase(query)) {
            return true;
        }
        String lower = key.toLowerCase(Locale.ROOT);
        return switch (query.toLowerCase(Locale.ROOT)) {
            case "project" -> lower.contains("project") || lower.contains("workspace");
            case "build" -> lower.contains("build") || lower.contains("gradle") || lower.contains("mvn");
            case "modules" -> lower.contains("module");
            case "run" -> lower.contains("run") || lower.contains("configuration");
            default -> true;
        };
    }

    private ObjectNode unsupportedBuildNode() {
        ObjectNode node = mapper.createObjectNode();
        node.put("ok", false);
        node.put("action", "build");
        node.put("status", "unsupported");
        node.put("code", "unsupported_operation");
        node.put("summary", "IDE build is disabled in cross-IDE mode. Use bash for build/test.");
        node.put("output", "IDE build is disabled in cross-IDE mode. Use bash for build/test.");
        return node;
    }

    private ObjectNode errorNode(String code, String message) {
        ObjectNode node = mapper.createObjectNode();
        node.put("ok", false);
        node.put("status", "error");
        node.put("code", code);
        node.put("summary", message);
        node.put("output", message);
        return node;
    }

    private JsonNode callBridgeJson(JsonNode payload) {
        IdeBridgeServer bridge;
        try {
            bridge = project.getService(IdeBridgeServer.class);
        } catch (Exception ignored) {
            bridge = null;
        }
        if (bridge == null) {
            return errorNode("bridge_unavailable", "IDE bridge unavailable. Ensure the plugin is running.");
        }
        String bridgeUrl = bridge.getBaseUrl();
        if (bridgeUrl == null || bridgeUrl.isBlank()) {
            return errorNode("bridge_not_started", "IDE bridge not started.");
        }
        try {
            String body = mapper.writeValueAsString(payload);
            HttpURLConnection connection = (HttpURLConnection) URI.create(bridgeUrl + "/execute").toURL().openConnection();
            connection.setRequestMethod("POST");
            connection.setRequestProperty("Content-Type", "application/json");
            connection.setDoOutput(true);
            connection.setConnectTimeout(5_000);
            connection.setReadTimeout(120_000);
            try (OutputStreamWriter writer = new OutputStreamWriter(connection.getOutputStream(), StandardCharsets.UTF_8)) {
                writer.write(body);
            }
            int status = connection.getResponseCode();
            byte[] responseBody = status >= 200 && status < 300
                    ? connection.getInputStream().readAllBytes()
                    : (connection.getErrorStream() == null ? new byte[0] : connection.getErrorStream().readAllBytes());
            connection.disconnect();
            if (responseBody.length == 0) {
                return errorNode("empty_response", "IDE bridge returned an empty response.");
            }
            String responseText = new String(responseBody, StandardCharsets.UTF_8);
            try {
                return mapper.readTree(responseText);
            } catch (Exception ignored) {
                return errorNode("invalid_response", "IDE bridge returned non-JSON response: " + responseText);
            }
        } catch (Exception e) {
            return errorNode("bridge_call_failed", "IDE bridge call failed: " + (e.getMessage() == null ? "unknown error" : e.getMessage()));
        }
    }

    private String prettyJson(JsonNode node) throws Exception {
        return mapper.writerWithDefaultPrettyPrinter().writeValueAsString(node);
    }

    private void copyIfPresent(JsonNode source, ObjectNode target, String... keys) {
        for (String key : keys) {
            if (source.has(key)) {
                JsonNode value = source.path(key);
                if (!value.isMissingNode() && !value.isNull()) {
                    target.set(key, value.deepCopy());
                }
            }
        }
    }

    private String firstNonBlank(String... values) {
        for (String value : values) {
            if (value != null && !value.trim().isBlank()) {
                return value.trim();
            }
        }
        return "";
    }

    private boolean isTerminalStatus(String status) {
        return List.of("succeeded", "success", "failed", "aborted", "canceled", "cancelled", "timeout").contains(status.trim().toLowerCase(Locale.ROOT));
    }

    public static class CapabilitySnapshot {
        private final boolean bridgeAvailable;
        private final List<String> actionOperations;

        public CapabilitySnapshot(boolean bridgeAvailable, List<String> actionOperations) {
            this.bridgeAvailable = bridgeAvailable;
            this.actionOperations = actionOperations == null ? List.of() : List.copyOf(actionOperations);
        }

        public boolean getBridgeAvailable() {
            return bridgeAvailable;
        }

        public List<String> getActionOperations() {
            return actionOperations;
        }
    }
}

File: idea-plugin/src/main/kotlin/com/zzf/rikki/idea/agent/tools/LiteIdeTools.kt
package com.zzf.rikki.idea.agent.tools

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.node.ObjectNode
import com.intellij.openapi.project.Project
import com.zzf.rikki.idea.IdeBridgeServer
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URI
import java.nio.charset.StandardCharsets

/**
 * IDE tools for the lite agent.
 *
 *  - ide_context     : returns context payload captured by ChatPanel
 *  - ide_action      : capability-gated bridge actions (run/test/status/cancel)
 *  - ide_capabilities: probes the current IDE bridge capabilities
 */
class LiteIdeTools(private val project: Project, private val mapper: ObjectMapper) {

    data class CapabilitySnapshot(
        val bridgeAvailable: Boolean = false,
        val actionOperations: List<String> = emptyList()
    )

    /** The latest ideContext payload received from ChatPanel, set by LiteAgentEngine before each run. */
    @Volatile var ideContextNode: JsonNode = mapper.createObjectNode()

    @Volatile private var cachedCapabilitySnapshot = CapabilitySnapshot()

    fun capabilitySnapshot(): CapabilitySnapshot = cachedCapabilitySnapshot

    fun refreshCapabilities(): CapabilitySnapshot {
        val node = callBridgeJson(mapper.createObjectNode().put("action", "capabilities"))
        val snapshot = parseCapabilitySnapshot(node)
        cachedCapabilitySnapshot = snapshot
        return snapshot
    }

    fun context(args: JsonNode): String {
        val ctx = ideContextNode
        if (ctx.isNull || ctx.isMissingNode || ctx.size() == 0) {
            return "No IDE context is available. Ensure the plugin is loaded and project is indexed."
        }

        val query = args.path("query").asText("all")
        val keys = buildList {
            if (args.has("keys") && args.path("keys").isArray) {
                args.path("keys").forEach { add(it.asText("")) }
            }
        }.filter { it.isNotBlank() }

        val filtered = linkedMapOf<String, JsonNode>()
        ctx.fields().forEach { (k, v) ->
            when {
                keys.isNotEmpty() && k in keys -> filtered[k] = v
                keys.isEmpty() && matchesQuery(k, query) -> filtered[k] = v
            }
        }

        return mapper.writerWithDefaultPrettyPrinter().writeValueAsString(filtered)
    }

    fun action(args: JsonNode): String {
        val operationRaw = firstNonBlank(args.path("operation").asText(""), args.path("action").asText("")).lowercase()
        if (operationRaw.isBlank()) {
            return prettyJson(errorNode("missing_operation", "Field 'operation' is required for ide_action."))
        }

        if (operationRaw == "capabilities") {
            return capabilities()
        }
        if (operationRaw == "build") {
            return prettyJson(unsupportedBuildNode())
        }

        val snapshot = refreshCapabilities()
        if (!snapshot.bridgeAvailable) {
            return prettyJson(errorNode("bridge_unavailable", "IDE bridge unavailable. Use bash for build/test operations."))
        }

        if (operationRaw !in snapshot.actionOperations) {
            return prettyJson(errorNode("unsupported_operation", "Unsupported ide_action operation: $operationRaw"))
        }

        val payload = mapper.createObjectNode()
        when (operationRaw) {
            "run", "test" -> {
                payload.put("action", "start")
                payload.put("operation", operationRaw)
            }
            "status" -> payload.put("action", "status")
            "cancel" -> payload.put("action", "cancel")
            else -> return prettyJson(errorNode("unsupported_operation", "Unsupported ide_action operation: $operationRaw"))
        }

        copyIfPresent(args, payload, "mode", "configuration", "configurationName", "name", "executor", "jobId", "sinceRevision", "waitMs")

        val firstResponse = callBridgeJson(payload)
        val wait = args.path("wait").asBoolean(false)
        if (!wait || operationRaw !in setOf("run", "test")) {
            return prettyJson(firstResponse)
        }

        val jobId = firstNonBlank(firstResponse.path("jobId").asText(""))
        if (!firstResponse.path("ok").asBoolean(false) || jobId.isBlank()) {
            return prettyJson(firstResponse)
        }

        if ("status" !in snapshot.actionOperations) {
            return prettyJson(firstResponse)
        }

        val timeoutMs = args.path("timeoutMs").asLong(120_000L).coerceIn(1_000L, 900_000L)
        val pollIntervalMs = args.path("pollIntervalMs").asLong(1_000L).coerceIn(200L, 10_000L)
        val deadline = System.currentTimeMillis() + timeoutMs
        var sinceRevision = 0L
        var latest: JsonNode = firstResponse

        while (System.currentTimeMillis() < deadline) {
            Thread.sleep(pollIntervalMs)
            val statusPayload = mapper.createObjectNode()
                .put("action", "status")
                .put("jobId", jobId)
                .put("sinceRevision", sinceRevision)
                .put("waitMs", pollIntervalMs)
            latest = callBridgeJson(statusPayload)
            sinceRevision = maxOf(sinceRevision, latest.path("logRevision").asLong(0L))
            if (isTerminalStatus(latest.path("status").asText(""))) {
                return prettyJson(latest)
            }
        }

        val timeoutNode = if (latest is ObjectNode) latest else mapper.createObjectNode()
        timeoutNode.put("ok", false)
        timeoutNode.put("status", "timeout")
        timeoutNode.put("summary", "Timed out waiting for IDE job completion.")
        timeoutNode.put("output", firstNonBlank(timeoutNode.path("output").asText(""), "Timed out waiting for IDE job completion."))
        return prettyJson(timeoutNode)
    }

    fun capabilities(): String {
        val node = callBridgeJson(mapper.createObjectNode().put("action", "capabilities"))
        cachedCapabilitySnapshot = parseCapabilitySnapshot(node)
        return prettyJson(node)
    }

    private fun parseCapabilitySnapshot(node: JsonNode): CapabilitySnapshot {
        if (!node.path("ok").asBoolean(false)) {
            return CapabilitySnapshot(false, emptyList())
        }
        val allowed = setOf("run", "test", "status", "cancel", "capabilities")
        val operations = linkedSetOf<String>()
        collectOps(node.path("asyncOperations"), operations, allowed)
        collectOps(node.path("directOperations"), operations, allowed)
        return CapabilitySnapshot(true, operations.toList())
    }

    private fun collectOps(source: JsonNode, dest: MutableSet<String>, allowed: Set<String>) {
        if (!source.isArray) return
        source.forEach { n ->
            val op = n.asText("").trim().lowercase()
            if (op in allowed) dest += op
        }
    }

    private fun matchesQuery(key: String, query: String): Boolean {
        if (query == "all") return true
        val k = key.lowercase()
        return when (query.lowercase()) {
            "project" -> k.contains("project") || k.contains("workspace")
            "build" -> k.contains("build") || k.contains("gradle") || k.contains("mvn")
            "modules" -> k.contains("module")
            "run" -> k.contains("run") || k.contains("configuration")
            else -> true
        }
    }

    private fun unsupportedBuildNode(): ObjectNode = mapper.createObjectNode().apply {
        put("ok", false)
        put("action", "build")
        put("status", "unsupported")
        put("code", "unsupported_operation")
        put("summary", "IDE build is disabled in cross-IDE mode. Use bash for build/test.")
        put("output", "IDE build is disabled in cross-IDE mode. Use bash for build/test.")
    }

    private fun errorNode(code: String, message: String): ObjectNode = mapper.createObjectNode().apply {
        put("ok", false)
        put("status", "error")
        put("code", code)
        put("summary", message)
        put("output", message)
    }

    private fun callBridgeJson(payload: JsonNode): JsonNode {
        val bridge = try {
            project.getService(IdeBridgeServer::class.java)
        } catch (_: Exception) { null }
            ?: return errorNode("bridge_unavailable", "IDE bridge unavailable. Ensure the plugin is running.")

        val bridgeUrl = bridge.getBaseUrl().ifBlank {
            return errorNode("bridge_not_started", "IDE bridge not started.")
        }

        return try {
            val body = mapper.writeValueAsString(payload)
            val conn = (URI.create("$bridgeUrl/execute").toURL().openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                setRequestProperty("Content-Type", "application/json")
                doOutput = true
                connectTimeout = 5_000
                readTimeout = 120_000
            }
            conn.outputStream.use { out ->
                OutputStreamWriter(out, StandardCharsets.UTF_8).use { it.write(body) }
            }
            val status = conn.responseCode
            val stream = if (status in 200..299) conn.inputStream else conn.errorStream
            val responseBody = stream?.use { it.readBytes().toString(StandardCharsets.UTF_8) } ?: ""
            conn.disconnect()

            if (responseBody.isBlank()) {
                return errorNode("empty_response", "IDE bridge returned an empty response.")
            }
            try {
                mapper.readTree(responseBody)
            } catch (_: Exception) {
                errorNode("invalid_response", "IDE bridge returned non-JSON response: $responseBody")
            }
        } catch (e: Exception) {
            errorNode("bridge_call_failed", "IDE bridge call failed: ${e.message ?: "unknown error"}")
        }
    }

    private fun prettyJson(node: JsonNode): String =
        mapper.writerWithDefaultPrettyPrinter().writeValueAsString(node)

    private fun firstNonBlank(vararg values: String): String {
        for (value in values) {
            val v = value.trim()
            if (v.isNotBlank()) return v
        }
        return ""
    }

    private fun copyIfPresent(source: JsonNode, target: ObjectNode, vararg keys: String) {
        for (k in keys) {
            if (!source.has(k)) continue
            val n = source.path(k)
            if (n.isMissingNode || n.isNull) continue
            target.set<JsonNode>(k, n.deepCopy())
        }
    }

    private fun isTerminalStatus(status: String): Boolean = when (status.trim().lowercase()) {
        "succeeded", "success", "failed", "aborted", "canceled", "cancelled", "timeout" -> true
        else -> false
    }
}
